src/zcl_abapgit_objects.clas.abap

```diff
@@ -591,33 +617,56 @@ CLASS ZCL_ABAPGIT_OBJECTS IMPLEMENTATION.
 
       li_obj->mo_files = lo_files.
 
-      IF li_obj->get_metadata( )-late_deser = abap_true.
-        APPEND INITIAL LINE TO lt_late ASSIGNING <ls_deser>.
-      ELSEIF li_obj->get_metadata( )-ddic = abap_true.
-        APPEND INITIAL LINE TO lt_ddic ASSIGNING <ls_deser>.
-      ELSE.
-        APPEND INITIAL LINE TO lt_rest ASSIGNING <ls_deser>.
-      ENDIF.
-      <ls_deser>-item    = ls_item.
-      <ls_deser>-obj     = li_obj.
-      <ls_deser>-xml     = lo_xml.
-      <ls_deser>-package = lv_package.
-
-      CLEAR: lv_path, lv_package.
+      TRY.
+* Get required steps for deserialize the object
+          lt_steps_id = li_obj->get_deserialize_steps( ).
+
+          LOOP AT lt_steps_id ASSIGNING <lv_step_id>.
+            READ TABLE lt_steps WITH KEY id = <lv_step_id> ASSIGNING <ls_step>.
+            IF sy-subrc <> 0.
+              zcx_abapgit_exception=>raise( |Step { <lv_step_id> } is not defined| ).
+            ELSEIF <ls_step>-is_ddic = abap_true AND li_obj->get_metadata( )-ddic = abap_false.
+              " DDIC only for DDIC obejcts
+              zcx_abapgit_exception=>raise( |Step { <lv_step_id> } is only for DDIC objects| ).
+            ENDIF.
+            APPEND INITIAL LINE TO <ls_step>-objects ASSIGNING <ls_deser>.
+            <ls_deser>-item    = ls_item.
+            <ls_deser>-obj     = li_obj.
+            <ls_deser>-xml     = lo_xml.
+            <ls_deser>-package = lv_package.
+          ENDLOOP.
+        CATCH cx_sy_dyn_call_illegal_method.
+          " Fallback (can be removed if all objects implement the new IF method get_deserialize_steps)
+          IF li_obj->get_metadata( )-late_deser = abap_true.
+            READ TABLE lt_steps WITH KEY id = gc_step_id-late ASSIGNING <ls_step>.
+            IF sy-subrc <> 0.
```

Comment:
suggest replacing these IF's with ASSERTs, as we know its there, its a program error if it fails

Reply:
fixed